### ISSUE TITLE
The calculation process is inconsistent with the original paper

### ISSUE THREAD

**Issue by chencodeX** [bionick87/ConvGRUCell-pytorch#1]
![image](https://user-images.githubusercontent.com/14937564/37450101-9d972d92-2867-11e8-9852-7e2a74674d2b.png)
In your implementation part.
In your implementation process, Equation (8) uses the following code:
        next_h       = torch.mul(update_gate,hidden) + (1-update_gate)*ct

According to the original calculation process, it should be the following code organization:
        next_h       = torch.mul(update_gate,ct) + (1-update_gate)*hidden

**Comment by lxtGH** [comment]
Yes, I agree with you. It is wrong

**Comment by bionick87** [comment]
Hi Guys,

I think is correct, I based my equation on this  blog:
a) https://esciencegroup.com/2016/03/04/fun-with-recurrent-neural-nets-one-more-dive-into-cntk-and-tensorflow/

How you can see  my output gate  is a function of gate vector ut  that is multiplied by h_{t-1} + (1-ut)*ct

Also you can find similar implementation here:

https://www.safaribooksonline.com/library/view/neural-networks-and/9781492037354/ch04.html

Now we try to understand way both implementation are similar. 

if you look in the original implementation h_t   is  *always* function of (h_(t-1),ct, update_gate)

But what does is mean?

The ut, update gate, control the information  that is getting in input to a h_{t-1} ( your previous hidden state).
that mean: how much information you want to inject in your previous state, plus the information that transit on ct ("your cell").

My implementation follow this original paper:

 https://arxiv.org/pdf/1406.1078v3.pdf
   
Is not wrong is just different way to update the gates but the principle is same: you want to control the    information on previous state and at the same time the information on the cell c with the same update gate!!!!

**Comment by lxtGH** [comment]
Thanks for reply,  I see, the standrad GRU is the same as your implementation in all other paper. But in **ConvGRU**(ICLR 2016) **paper**, formula 4 of GRU is **different** from all other paper. Is this a mistake or they are the same with each other? I am not quiet familar with RNN.

**Comment by bionick87** [comment]
Hi, 

They are equivalent and symmetrical.

You can see it as a linear interpolation between two states: the recurrent state and the candidate cell state OR in the *other way around*. 

Then, the equations are both symmetric.

Another interpretation of this (1- ut) could be linked to the prevention of the gradient explosion.
In fact, if we imagine that the gradient explodes the *update gate* is saturated to 1.
Then it cancel OR the cell term ct OR h_{t-1}. But in this limit case, it makes no difference to keep the recurring state or the memory cell, the system has suffered an explosion!

Best,

Nicolò
